checks.d/docker_daemon.py

```diff
@@ -0,0 +1,640 @@
+                            tags.append('%s:%s' % (tag_name, str(t).strip()))
+
+            # Add ECS tags
+            if self.collect_ecs_tags:
+                entity_id = entity.get("Id")
+                if entity_id in self.ecs_tags:
+                    ecs_tags = self.ecs_tags[entity_id]
+                    tags.extend(ecs_tags)
+
+        return tags
+
+    def _extract_tag_value(self, entity, tag_name):
+        """Extra tag information from the API result (containers or images).
+        Cache extracted tags inside the entity object.
+        """
+        if tag_name not in TAG_EXTRACTORS:
+            self.warning("{0} isn't a supported tag".format(tag_name))
+            return
+
+        # Check for already extracted tags
+        if "_tag_values" not in entity:
+            entity["_tag_values"] = {}
+
+        if tag_name not in entity["_tag_values"]:
+            entity["_tag_values"][tag_name] = TAG_EXTRACTORS[tag_name](entity)
+
+        return entity["_tag_values"][tag_name]
+
+    def refresh_ecs_tags(self):
+        ecs_config = self.client.inspect_container('ecs-agent')
+        net_conf = ecs_config['NetworkSettings'].get('Ports', {})
+        net_conf = net_conf.get(net_conf.keys()[0], [])
+        ecs_tags = {}
+        if net_conf:
+            net_conf = net_conf[0] if isinstance(net_conf, list) else net_conf
+            ip, port = net_conf.get('HostIp'), net_conf.get('HostPort')
+            tasks = requests.get('http://%s:%s' % (ip, port)).json()
+            for task in tasks.get('Tasks', []):
+                for container in task.get('Containers', []):
+                    tags = ['task_name:%s' % task['Family'], 'task_version:%s' % task['Version']]
+                    ecs_tags[container['DockerId']] = tags
+
+        self.ecs_tags = ecs_tags
+
+    def _filter_containers(self, containers):
+        if not self._filtering_enabled:
+            return
+
+        for container in containers:
+            container_tags = self._get_tags(container, FILTERED)
+            container['_is_filtered'] = self._are_tags_filtered(container_tags)
+
+            if container['_is_filtered']:
+                self.log.debug("Container {0} is filtered".format(container["Names"][0]))
+
+    def _are_tags_filtered(self, tags):
+        if self._tags_match_patterns(tags, self._exclude_patterns):
+            if self._tags_match_patterns(tags, self._include_patterns):
+                return False
+            return True
+        return False
+
+    def _tags_match_patterns(self, tags, filters):
+        for rule in filters:
+            for tag in tags:
+                if re.match(rule, tag):
+                    return True
+        return False
+
+    def _is_container_excluded(self, container):
+        """Check if a container is excluded according to the filter rules.
+
+        Requires _filter_containers to run first.
+        """
+        return container.get('_is_filtered', False)
```

Comment:
I'm not sure I like that `_get_and_count_containers` is opaquely mutating the container dict in a way that subsequent functions rely on (e.g. `_report_container_size`). Can we have `_is_container_excluded` actually call `_filter_containers` if the key is missing, and hence always return an accurate value? It would make debugging checks a lot easier imo, not to mention that it makes the function name actually truthful